### DISCUSSION TITLE
Mixin Relation Not Using Instance id in primary join

### DISCUSSION THREAD

**Discussion by radiantone** [sqlalchemy/sqlalchemy#7073] [Usage Questions]
Hi,
  I have this mixin relation with a many-to-one relationship using a single table with a column "oid" in LogModel that maps to the "id" column of the class INSTANCE it's mixed in with. Creating an instance of Processor and adding Log objects to it works fine. And LogModel table has the rows and oid reference in it.
  
  Problem is, the primary join is just grabbing whatever rows in the remote table whose "id" can be joined and thus I get the same log rows for all objects. So no matter what Processor object I have retrieved, it always fetches the same rows (for another processor).

How can I constrain the primary join to force the use of the "id" of the class "instance" it's mixed in with and not select any matching id from the same table?

```
class HasLogs(object):

    @declared_attr
    def logs(cls):
        return relationship(
            "LogModel",
            order_by="desc(LogModel.created)",
            primaryjoin=lambda: foreign(LogModel.oid) == remote(cls.id),   # Here. cls.id is always satisfied by ANY row
                                                                               # I want cls.id to match the actual object when I access
                                                                              # processor.logs in my code.
            lazy="select"
        )
        
class Processor(HasLogs, BaseModel):
    ...
```

**Comment by zzzeek** [comment]
this is still #7064, so now that we are here, lets stay here until this model you're doing is everything you want.

relationally, you have this:

```
procesors
-------------
id

some_other_thing
-----------------------
id


log_table
------------

oid -> foreign key to either processors(id) or some_other_thing(id)
```

from a relational algebra perspective, the above design is wrong.  you can put garbage data into log_table.oid that points nowhere, there is no foreign key constraint that can detect it, and now your database is corrupted.  

However, as I mentioned, this pattern is a popular one that is used by django and ruby on rails.   Our example of this pattern is here: https://docs.sqlalchemy.org/en/14/_modules/examples/generic_associations/generic_fk.html

what's important in that pattern is another column called "discriminator".  this is an often textual column ( or enum, integer, whatever) that refers to the type of object we are pointing towards.

Therefore you need to add a column to Log called "discriminator".  Then set up the primary join as you see in the example, like this:

```py
class HasLogs(object):

    @declared_attr
    def logs(cls):
        return relationship(
            "LogModel",
            order_by="desc(LogModel.created)",
            primaryjoin=lambda: and_(foreign(LogModel.oid) == remote(cls.id),   LogModel.discriminator == cls.__name__)
            lazy="select"
        )
```

above it uses "class name", in this case "Processor" as the discriminator value.  you can have it set up as other things too.

the only thing you have to do now is make sure when you create these Log objects that you also set the correct "discriminator" vlaue on them.  the example does this using the "append" event, but you could also do it using things like the [before_insert](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_insert#sqlalchemy.orm.MapperEvents.before_insert) event.  or most simply you would do it manually:    `Processor.logs.append(LogEvent(discriminator="Processor"))`.    

Still another way is to use [single table inheritance](https://docs.sqlalchemy.org/en/14/orm/inheritance.html#single-table-inheritance) to set up these discriminators more clearly; you'd have subclasses like ProcessorLog(Log), SomeOtherLog(Log), etc., but they all get stored in that one "logs" table.  you can query for them directly and the ORM will also automatically add the "discriminator" to ORM queries, you wouldn't need to add the "discriminator" to the primaryjoin condition.

the basic pattern you see here is called "generic foreign key", compare to django's pattern at https://docs.djangoproject.com/en/1.10/ref/contrib/contenttypes/#generic-relations along with an old critique: https://lukeplant.me.uk/blog/posts/avoid-django-genericforeignkey/  however I'd note SQLAlchemy doesn't have the performance problems mentioned there.  just the referential integrity part.

**Reply by radiantone** [reply]
I see your point about the lack of ForeignKey on LogModel. The tricky part there is (as mentioned in other thread), indicating a foreignkey a priori of the instance actually using it. In the other thread, that solution ran into a separate problem whereby the mixin inheriting from (object) did not have its table built and pulling objects from the database resulted in table not found when it tried to access that relation (logs). Unless it already had logs for it. Then it worked. But separate problem.

Back to this one. I will try adding the discriminator, but the part I struggle to see here is how that will select one particular Processor instance vs another? So to elaborate the problem I ran into with my design here.

1. I created a ProcessorModel object
2. I created 2 LogModel objects and added them to the ProcessorModel. Works fine. Now I have two log entry in "logs" table with oid pointing to ProcessorModel
```
    
    log1 = LogModel(oid=processor.id, text='This is a log', source='this')
    log2 = LogModel(oid=processor.id, text='This is a log too', source='that')
    
    processor.logs += [log1]
    processor.logs += [log2]
 ```
3. I search for ProcessorModel in the database and get handle to it, and processor.logs contains the two logs. Great!
4. Created a second ProcessorModel object with different id. When I try to access .logs of that model, it comes back with the logs from the first processor, even though the oid does not match the id of the second processor.
 
So by adding a Discriminator. I'm not seeing how this gets resolved since both Processor objects (1 and 2) are of the same class.
Seems like the primary join still needs to use the instance of processor.id to build the select statement for the join and its not doing that out-of-the-box.

**Reply by radiantone** [reply]
Just to reiterate the behavior I'm looking for (for onlookers).

1. I have a bunch of model objects that inhereit from BaseModel. Which provide columns like id, created, name, owner etc.
2. I want to add a one-to-many relation to all objects for "logs" such that each model now has a "logs" collection of LogModel objects.
3. All the LogModels live in one table so easy query and reporting can be done without knowing the type of parent class of the log at query time.
4. Use a mixin class to add the relationship
5. Do not want specific Log classes for each object in my database.
6. Do not want association tables. Creates too many tables and complicates queries and reporting.
7. Would prefer just adding the mixin class to the abstract base class rather than editing each model, but not a huge requirement.

**Reply by radiantone** [reply]
@zzzeek 
Hi Mike,
  I tried the discriminator and it didn't work. Same result, 2 different ProcessorModels ( A & B where A has 2 logs and B has none). Both return ProcessorModel A's logs. This is because the declared_attr doesn't have access to the instance (A & B's id), only the class. 
  
 Yet, all the data is there to retrieve the log rows based on the one-to-many relation. So probably this is a gap in SQLAlchemy design. 
 
 Consider my data:
 
2 Log rows with oid pointing to parent ProcessorModel. In this case bfb..... 
![image](https://user-images.githubusercontent.com/25509097/134799073-5b1b50e2-425c-4511-b095-8ee172760490.png)

And here, 2 processor model rows. One with id = 'bfb.....' that matches the oid of the log rows
![image](https://user-images.githubusercontent.com/25509097/134799121-45814fc8-279c-4f19-ae18-4eccee73fe2c.png)

And my mixin. See where it says cls.id? That should translate to whatever instance I am accessing the logs relation for (intuitively), but it doesn't.
```
class HasLogs(object):

    @declared_attr
    def logs(cls):
        return relationship(
            "LogModel",
            order_by="desc(LogModel.created)",
            primaryjoin=lambda: and_(foreign(LogModel.oid) == remote(cls.id),   LogModel.discriminator == cls.__name__),
            lazy="select"
        )
``` 
These produce the same rows, wrong oid==id match
```
proc1 = query for ProcessorModel with name 'proc1' # The correct parent
for log in proc1.logs: print(log)    # shows correct logs

proc2 = query for ProcessorModel with name 'proc2' # Should have no logs
for log in proc2.logs: print(log)    # shows logs for proc1 
```
This is because the select/join is wrong. It is using any satisfying rows from both tables, when it should constrain 
```
log.oid = `bfbf21cf-56b0-43f1-89cd-b948da32faf5` # For whatever the processor object being used is
```
not
```
log.oid = processor.id
```

**Reply by zzzeek** [reply]
hi - 

at this point I can only work with self-contained examples that show a failed assertion.  take the self-contained example I wrote at https://github.com/sqlalchemy/sqlalchemy/discussions/7064#discussioncomment-1366944 and fill in your current model design, then a few rows of sample data.  then show me the query that does not return what's expected.  thanks

**Reply by radiantone** [reply]
Ok. I will provide a self contained example. Thank you

**Reply by radiantone** [reply]
@zzzeek 
Hey Mike, 
  Here is a reproducible test case that shows the problem I explained above, which is the primary join is not using the absolute value for the cls.id object. 
  
  ```
  
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import declarative_mixin, foreign, remote
from sqlalchemy.ext.declarative import DeclarativeMeta
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.schema import CreateColumn
from sqlalchemy import Enum, Table, Column, Integer, LargeBinary, Text, String, ForeignKey, DateTime, Boolean, Float, Sequence, INTEGER, literal_column, select, column
from sqlalchemy.dialects.postgresql import DOUBLE_PRECISION

from datetime import datetime

from sqlalchemy.orm import relationship
from sqlalchemy.orm import declarative_base
from sqlalchemy import create_engine
from sqlalchemy import and_
from sqlalchemy.orm import sessionmaker

import platform

HOSTNAME = platform.node()

# Edit your password in here
POSTGRES_ROOT = 'postgresql://postgres:PASSWORD@'+HOSTNAME+':5432/'

Base = declarative_base(name="Base")

class HasLogs(object):

    @declared_attr
    def logs(cls):
        return relationship(
            "LogModel",
            order_by="desc(LogModel.created)",
            primaryjoin=lambda: and_(foreign(LogModel.oid) == remote(
                cls.id),   LogModel.discriminator == cls.__name__),
            lazy="select"
        )


class BaseModel(Base):
    __abstract__ = True

    id = Column(String(40), autoincrement=False, default=literal_column(
        'uuid_generate_v4()'), unique=True, primary_key=True)
    name = Column(String(80), unique=True, nullable=False, primary_key=True)
    owner = Column(String(40), default=literal_column('current_user'))
    created = Column(DateTime, default=datetime.now, nullable=False)
    lastupdated = Column(DateTime, default=datetime.now,
                         onupdate=datetime.now, nullable=False)


class LogModel(Base):
    __tablename__ = 'log'

    id = Column(String(40), autoincrement=False, default=literal_column(
        'uuid_generate_v4()'), unique=True, primary_key=True)

    created = Column(DateTime, default=datetime.now, nullable=False)
    oid = Column(String(40), primary_key=True)
    discriminator = Column(String(40))
    text = Column(String(80), nullable=False)
    source = Column(String(40), nullable=False)

    def __repr__(self):
        return "{}:{}:{}".format(self.id, self.source, self.text)


class ProcessorModel(HasLogs, BaseModel):
    __tablename__ = 'processor'

    hostname = Column(String(60))

    def __repr__(self):
        return '{}:{}:{}:{}:{}:{}:{}:{}:{} Plugs:{} Sockets:{}'.format(self.id, self.name, self.beat, self.lastupdated, self.hostname, self.concurrency, self.requested_status, self.status, self.worker, self.plugs, self.sockets)


def init_db():
    from sqlalchemy import create_engine
    engine = create_engine(POSTGRES_ROOT+'postgres')
    session = sessionmaker(bind=engine)()
    session.connection().connection.set_isolation_level(0)
    session.execute('CREATE DATABASE testdb')
    session.connection().connection.set_isolation_level(1)
    engine = create_engine(POSTGRES_ROOT + 'testdb')
    session = sessionmaker(bind=engine)()
    session.execute('CREATE EXTENSION IF NOT EXISTS "uuid-ossp"')
    engine.session = session
    Base.metadata.create_all(engine)
    print("Created database.")
    return session

def test(session):
    proc1 = ProcessorModel(name="proc1")
    proc2 = ProcessorModel(name="proc2")
    log1 = LogModel(oid=proc1.id, text='This is a log',
                    discriminator='ProcessorModel', source='test')
    log2 = LogModel(oid=proc1.id, text='This is a log too',
                    discriminator='ProcessorModel', source='test')
    proc1.logs += [log1]
    proc1.logs += [log2]
    session.add(proc1)
    session.add(proc2)
    session.add(log1)
    session.add(log2)
    session.commit()

def search(session):
    proc1 = session.query(ProcessorModel).filter_by(name="proc1").first()
    print("Proc1 logs:")
    for log in proc1.logs:
        print(log)

    proc2 = session.query(ProcessorModel).filter_by(name="proc2").first()
    print("Proc2 logs:")
    for log in proc2.logs:
        print(log)

session = init_db()
test(session)
search(session)
```
Produces output:
```
$ python logtest.py 
Created database.
Proc1 logs:
b48fe10c-aea8-4ab6-a356-c6f1d8f70bf9:test:This is a log too
a949f876-14e0-4fbb-97ab-5849d6c69bdf:test:This is a log
Proc2 logs:
b48fe10c-aea8-4ab6-a356-c6f1d8f70bf9:test:This is a log too
a949f876-14e0-4fbb-97ab-5849d6c69bdf:test:This is a log
```
Which shows the same logs for both ProcessorModels even though they are only added to proc1.

**Reply by zzzeek** [reply]
you need to fix that primaryjoin, should be like this:

```
            primaryjoin=lambda: and_(
                foreign(LogModel.oid) == cls.id,
                LogModel.discriminator == cls.__name__,
            ),
```

"cls.id" is not "remote", it's local.  You never need to set "remote" on a non-self-referential relationship, it is automatically the column that's on the table being pointed towards, in this case LogModel.

turn on echo=True on the engine, the query at the end changes from 

```
SELECT log.id AS log_id, log.created AS log_created, log.oid AS log_oid, log.discriminator AS log_discriminator, log.text AS log_text, log.source AS log_source 
FROM log, processor 
WHERE log.oid = processor.id AND log.discriminator = %(discriminator_1)s ORDER BY log.created DESC
2021-09-26 21:29:12,592 INFO sqlalchemy.engine.Engine [cached since 0.002786s ago] {'discriminator_1': 'ProcessorModel'}
```

to 

```
SELECT log.id AS log_id, log.created AS log_created, log.oid AS log_oid, log.discriminator AS log_discriminator, log.text AS log_text, log.source AS log_source 
FROM log 
WHERE log.oid = %(param_1)s AND log.discriminator = %(discriminator_1)s ORDER BY log.created DESC
2021-09-26 21:29:35,362 INFO sqlalchemy.engine.Engine [cached since 0.004029s ago] {'param_1': '661cc336-f713-49a8-9c2c-5b020d68ca95', 'discriminator_1': 'ProcessorModel'}
```

the fact that "processor.id" is not "remote" allows the lazy loader to know where it needs to fill in the primary key of the local object.

I'm not sure why we haven't done this before but I think technically there should be a warning when remote() is applied to a column on the local table, that isn't present in any of the referenced tables.   just checked to see if we have an issue for that but apparently not.

**Reply by zzzeek** [reply]
#7094

**Reply by radiantone** [reply]
It worked. Thank you.
